src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+          {
+            var parms = item as TestCaseParameters;
+
+            if (parms == null)
+            {
+              object[] args = item as object[];
+              if (args != null)
+              {
+                #if NETCF
+                if (method.IsGenericMethodDefinition)
+                {
+                  var mi = method.MakeGenericMethodEx(args);
+                  parameters = mi == null ? new ParameterInfo[0] : mi.GetParameters();
+                }
+                #endif
+                if (args.Length != parameters.Length)
+                  args = new object[] { item };
+              }
+              else if (item is Array)
+              {
+                Array array = item as Array;
+
+                #if NETCF
+                if (array.Rank == 1 && (method.IsGenericMethodDefinition || array.Length == parameters.Length))
+                  #else
+                  if (array.Rank == 1 && array.Length == parameters.Length)
+                    #endif
+                {
+                  args = new object[array.Length];
+                  for (int i = 0; i < array.Length; i++)
+                    args[i] = array.GetValue(i);
+                  #if NETCF
+                  if (method.IsGenericMethodDefinition)
+                  {
+                    var mi = method.MakeGenericMethodEx(args);
+
+                    if (mi == null || array.Length != mi.GetParameters().Length)
+                      args = new object[] {item};
+                  }
+                  #endif
+                }
+                else
+                {
+                  args = new object[] { item };
+                }
+              }
+              else
+              {
+                args = new object[] { item };
+              }
+
+              parms = new TestCaseParameters(args);
+            }
+
+            data.Add(parms);
+          }
+        }
+      }
+      catch (Exception ex)
+      {
+        data.Clear();
+        data.Add(new TestCaseParameters(ex));
+      }
+
+      return data;
+    }
+    
+    #endregion
+    
+    #region ITestBuilder Members
+
+    /// <summary>
+    /// Construct one or more TestMethods from a given MethodInfo,
+    /// using available parameter data.
+    /// </summary>
+    /// <param name="method">The MethodInfo for which tests are to be constructed.</param>
+    /// <param name="suite">The suite to which the tests will be added.</param>
+    /// <returns>One or more TestMethods</returns>
+    public IEnumerable<TestMethod> BuildFrom(IMethodInfo method, Test suite)
+    {
+      foreach (TestCaseParameters tcp in GetTestCasesFor(method))
+        yield return _builder.BuildTestMethod(method, suite, tcp);
+    }
+    
+    #endregion
+
+    #region Helper Methods
+
+    private IEnumerable GetDataFromSource(IMethodInfo method){
+      
+      #region Error Checking
+      
+      // file name sources
+      if (DataType == DataSourceType.Csv || DataType == DataSourceType.Xml){
```

Comment:
I'd probably consider your use case to be a CSV with a single column. If I separate the source types to individual classes, could we not differentiate by type instead of the property?